crates/ruff_python_formatter/src/expression/binary_like.rs

```diff
@@ -20,13 +20,182 @@ use crate::expression::string::StringLayout;
 use crate::expression::OperatorPrecedence;
 use crate::prelude::*;
 
-pub(super) struct BinaryLike<'a>(pub(super) &'a ExprBinOp);
+#[derive(Copy, Clone, Debug)]
+pub(super) enum BinaryLike<'a> {
+    BinaryExpression(&'a ExprBinOp),
+    CompareExpression(&'a ExprCompare),
+}
+
+impl<'a> BinaryLike<'a> {
+    fn flatten(
+        self,
+        comments: &'a Comments<'a>,
+        source: &str,
+    ) -> FormatResult<FlatBinaryExpression<'a>> {
+        fn recurse_compare<'a>(
+            compare: &'a ExprCompare,
+            leading_comments: &'a [SourceComment],
+            trailing_comments: &'a [SourceComment],
+            comments: &'a Comments,
+            source: &str,
+            parts: &mut SmallVec<[OperandOrOperator<'a>; 8]>,
+        ) -> FormatResult<()> {
+            parts.reserve(compare.comparators.len() * 2 + 1);
+
+            rec(
+                Operand::Left {
+                    expression: &compare.left,
+                    leading_comments,
+                },
+                comments,
+                source,
+                parts,
+            )?;
+
+            if compare.comparators.len() != compare.ops.len() {
+                return Err(FormatError::syntax_error(
+                    "Compare expression with an unbalanced number of comparators and operations.",
+                ));
+            }
```

Comment:
This should never happen for a valid tree but using a result (for which the formatter has good infrastructure) allows me to avoid the panic (and the resulting ruff crash). This is also similar to other code where we return a SyntaxError when we e.g. fail to find a keyword. 

But I don't feel strongly. We can also convert this to an assertion (I don't want to use a debug assertion because I want to abort to avoid formatting broken code and potentially removing code)

Reply:
my rule of thumb is: If i can't write a reasonable test case where this error gets returned, i use an assertion. I think a crash is fine than because we're then in a state we should never be in (and recovery is not modifying the user's file)